plugins/hypervisors/kvm/src/main/java/org/apache/cloudstack/utils/linux/KVMHostInfo.java

```diff
@@ -0,0 +1,132 @@
+
+import com.cloud.hypervisor.kvm.resource.LibvirtCapXMLParser;
+import com.cloud.hypervisor.kvm.resource.LibvirtConnection;
+import org.apache.commons.io.IOUtils;
+import org.apache.log4j.Logger;
+import org.libvirt.Connect;
+import org.libvirt.LibvirtException;
+import org.libvirt.NodeInfo;
+
+import java.io.FileReader;
+import java.io.IOException;
+import java.io.Reader;
+import java.util.ArrayList;
+import java.util.List;
+
+public class KVMHostInfo {
+
+    private static final Logger LOGGER = Logger.getLogger(KVMHostInfo.class);
+
+    private int cpus;
+    private int cpusockets;
+    private long cpuSpeed;
+    private long totalMemory;
+    private long reservedMemory;
+    private long overCommitMemory;
+    private List<String> capabilities = new ArrayList<>();
+
+    public KVMHostInfo(long reservedMemory, long overCommitMemory) {
+        this.reservedMemory = reservedMemory;
+        this.overCommitMemory = overCommitMemory;
+        this.getHostInfoFromLibvirt();
+        this.totalMemory = new MemStat(this.getReservedMemory(), this.getOverCommitMemory()).getTotal();
+    }
+
+    public int getCpus() {
+        return this.cpus;
+    }
+
+    public int getCpuSockets() {
+        return this.cpusockets;
+    }
+
+    public long getCpuSpeed() {
+        return this.cpuSpeed;
+    }
+
+    public long getTotalMemory() {
+        return this.totalMemory;
+    }
+
+    public long getReservedMemory() {
+        return this.reservedMemory;
+    }
+
+    public long getOverCommitMemory() {
+        return this.overCommitMemory;
+    }
+
+    public List<String> getCapabilities() {
+        return this.capabilities;
+    }
+
+    protected static long getCpuSpeed(final NodeInfo nodeInfo) {
+        try (final Reader reader = new FileReader(
+                "/sys/devices/system/cpu/cpu0/cpufreq/cpuinfo_max_freq")) {
+            return Long.parseLong(IOUtils.toString(reader).trim()) / 1000;
+        } catch (IOException | NumberFormatException e) {
+            LOGGER.info("Could not read cpuinfo_max_freq, falling back on libvirt");
+            return nodeInfo.mhz;
+        }
+    }
+
+    private void getHostInfoFromLibvirt() {
+        try {
+            final Connect conn = LibvirtConnection.getConnection();
+            final NodeInfo hosts = conn.nodeInfo();
+            this.cpuSpeed = getCpuSpeed(hosts);
+
+            /*
+             * Some CPUs report a single socket and multiple NUMA cells.
+             * We need to multiply them to get the correct socket count.
+             */
+            this.cpusockets = hosts.sockets;
+            if (hosts.nodes > 0) {
+                this.cpusockets = hosts.sockets * hosts.nodes;
+            }
+            this.cpus = hosts.cpus;
+
+            final LibvirtCapXMLParser parser = new LibvirtCapXMLParser();
+            parser.parseCapabilitiesXML(conn.getCapabilities());
+            final ArrayList<String> oss = parser.getGuestOsType();
+            for (final String s : oss) {
```

Comment:
Why doing all of this process if you only report the host supporting HVM? I mean, KVM does not run if the CPU does not support HVM. Therefore, all KVM hosts will always have HVM support, right?

Reply:
Yes, they will. But what I did is take the original code and use it. I didn't modify the logic behind the capabilities.

I would make that a different PR if we want to do that.

Reply:
I understand that you are only moving the old code around, but why not remove useless code now?

I mean, why would we keep a code that does not aggregate value, and can be misleading to newcomers?